syntax = "proto3";

package desktop;

//import "google/protobuf/any.proto";
 
service Desktop {
    rpc List(ListRequest) returns (ListResponse) {}
    rpc ListState(ListStateRequest) returns (ListStateResponse) {}
    rpc Get(GetRequest) returns (GetResponse) {}
    rpc GetState(GetStateRequest) returns (GetStateResponse) {}
    rpc FromTemplate(FromTemplateRequest) returns (FromTemplateResponse) {}
    rpc FromMedia(FromMediaRequest) returns (FromMediaResponse) {}
    rpc Update(UpdateRequest) returns (UpdateResponse) {}
    rpc Delete(DeleteRequest) returns (DeleteResponse) {}
    rpc Start(StartRequest) returns (StartResponse) {}
    rpc Stop(StopRequest) returns (StopResponse) {}
    rpc Viewer(ViewerRequest) returns (ViewerResponse) {}
}

enum Action {
    UNKNOWN = 0;
    STOP = 1;
    START = 2;
    PAUSE = 3;
    RESUME = 4;
    DELETE = 5;
    UPDATE = 6;
    TEMPLATE = 7;
}

message ListRequest {}

message ListResponse {
    message DesktopFields {
        string desktop_id = 1;
        Hardware hardware = 2;
    }
    repeated DesktopFields desktops = 1;
}

message ListStateRequest {}

message DesktopState {
    enum State {
        UNKNOWN = 0;
        STOPPED = 1;
        STARTED = 2;
        PAUSED = 3;
        DELETED = 4;
        FAILED = 5;
    }
    string desktop_id = 1;
    State state = 2;
    repeated Action next_actions = 3;
}

message ListStateResponse {
    repeated DesktopState desktops = 1;
}

message GetRequest {
    string desktop_id = 1;
}

message GetResponse {
    enum State {
        UNKNOWN = 0;
        STOPPED = 1;
        STARTED = 2;
        PAUSED = 3;
        DELETED = 4;
        FAILED = 5;
    }
    Hardware hardware = 1;
    State state = 2;
    repeated Action next_actions = 3;
}

message GetStateRequest {
    string desktop_id = 1;
}

message GetStateResponse {
    enum State {
        UNKNOWN = 0;
        STOPPED = 1;
        STARTED = 2;
        PAUSED = 3;
        DELETED = 4;
        FAILED = 5;
    }
    State state = 1;
    repeated Action next_actions = 3;
}

message FromTemplateRequest {
    string desktop_id = 1;
    string template_id = 2;
    Hardware hardware = 3;
}

message FromTemplateResponse {
    enum State {
        UNKNOWN = 0;
        STOPPED = 1;
        FAILED = 2;
    }

    State state = 1;
    repeated Action next_actions = 2;
}

message FromMediaRequest {
    string desktop_id = 1;
    string virt_install_id = 2;
    Hardware hardware = 3;
}

message FromMediaResponse {
    enum State {
        UNKNOWN = 0;
        STOPPED = 1;
        FAILED = 2;
    }
    State state = 1;
    repeated Action next_actions = 3;
}

message UpdateRequest {
    string desktop_id = 1;
    HardwareUpdate hardware = 2;
}

message UpdateResponse {
    enum State {
        UNKNOWN = 0;
        STOPPED = 1;
        FAILED = 2;
    }
    State state = 1;
    repeated Action next_actions = 2;
}

message DeleteRequest {
    string desktop_id = 1;
}

message DeleteResponse {
    enum State {
        UNKNOWN = 0;
        DELETED = 1;
        FAILED = 2;
    }
    State state = 1;
    repeated Action next_actions = 2;
}

message StartRequest {
    string desktop_id = 1;
}

message StartResponse {
    enum State {
        UNKNOWN = 0;
        STOPPED = 1;
        STARTED = 2;
        PAUSED = 3;
        DELETED = 4;
        FAILED = 5;
    }
    State state = 1;
    Viewer viewer = 2;
    repeated Action next_actions = 3;
}

message StopRequest {
    string desktop_id = 1;
}

message StopResponse {
    enum State {
        UNKNOWN = 0;
        STOPPED = 1;
        STARTED = 2;
        PAUSED = 3;
        DELETED = 4;
        FAILED = 5;
    }
    State state = 1;
    repeated Action next_actions = 2;
}

message ViewerRequest {
    string desktop_id = 1;
}

message ViewerResponse {
    Viewer viewer = 1;
    repeated Action next_actions = 2;
}

message Viewer {
    string hostname = 1;
    string hostname_external = 2;
    int32 port_spice = 5;
    int32 port_spice_ssl = 6;
    int32 port_vnc = 7;
    int32 port_vnc_websocket = 8;
    string passwd = 9;
    string client_addr = 10;
    float client_since = 11;
}

message Hardware {
    int32 vcpus = 1;
    int32 memory = 2;
    string boot_disk_rpath = 3;
    string boot_disk_bus = 4;
    int32  boot_disk_size = 5; 
    repeated string videos = 6;
    repeated string graphics = 7;
    repeated string boots = 8;
    repeated string interfaces = 9;
    repeated string isos = 10;
    repeated string floppies = 11;
}

message HardwareDerived {
    int32 vcpus = 1;
    int32 memory = 2;
    string boot_disk_rpath = 3;
    string boot_disk_bus = 4; 
    repeated string videos = 5;
    repeated string graphics = 6;
    repeated string boots = 7;
    repeated string interfaces = 8;
    repeated string isos = 9;
    repeated string floppies = 10;
}

message HardwareUpdate {
    int32 vcpus = 1;
    int32 memory = 2;
    string boot_disk_bus = 7; 
    repeated string videos = 3;
    repeated string graphics= 4 ;
    repeated string boots = 5;
    repeated string interfaces = 6;
    repeated string isos = 8;
    repeated string floppies = 9;
}


